File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        // if(head==null || head.next==null){
        //     return head;
        // }
        ListNode f=head;
        ListNode s=head;
        while(f!=null && f.next != null){
            f=f.next.next;
            s=s.next;
            if(f==s){
                while(head != s){
                    head=head.next;
                    s=s.next;
                }
                return s;
            }
        }
            return null;

    }
}

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode DetectCycle(ListNode head) {
        var set = new HashSet<ListNode>();
        
        while (head != null)
        {
            if (set.Contains(head))
                break;
            
            set.Add(head);
            head = head.next;
        }

        return head;
    }
}